File: java/ReverseLinkedList.java
/**
 * Reverse Linked List Problem Solution
 * 
 * Problem: Given the head of a singly linked list, reverse the list, 
 * and return the reversed list.
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class ReverseLinkedList {
    
    public static ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode current = head;
        
        while (current != null) {
            ListNode nextTemp = current.next;
            current.next = prev;
            prev = current;
            current = nextTemp;
        }
        
        return prev;
    }
    
    // Helper method to create linked list from array
    public static ListNode createLinkedList(int[] arr) {
        if (arr.length == 0) return null;
        
        ListNode head = new ListNode(arr[0]);
        ListNode current = head;
        
        for (int i = 1; i < arr.length; i++) {
            current.next = new ListNode(arr[i]);
            current = current.next;
        }
        
        return head;
    }
    
    // Helper method to print linked list
    public static void printLinkedList(ListNode head) {
        ListNode current = head;
        System.out.print("[");
        while (current != null) {
            System.out.print(current.val);
            if (current.next != null) System.out.print(",");
            current = current.next;
        }
        System.out.println("]");
    }
    
    // Helper method to convert linked list to array
    public static int[] linkedListToArray(ListNode head) {
        if (head == null) return new int[0];
        
        // Count nodes
        int count = 0;
        ListNode current = head;
        while (current != null) {
            count++;
            current = current.next;
        }
        
        // Convert to array
        int[] arr = new int[count];
        current = head;
        int i = 0;
        while (current != null) {
            arr[i++] = current.val;
            current = current.next;
        }
        
        return arr;
    }
    
    // Main method for testing
    public static void main(String[] args) {
        // Test case 1
        int[] arr1 = {1, 2, 3, 4, 5};
        ListNode head1 = createLinkedList(arr1);
        System.out.println("Original list:");
        printLinkedList(head1);
        
        ListNode reversed1 = reverseList(head1);
        System.out.println("Reversed list:");
        printLinkedList(reversed1);
        
        // Test case 2
        int[] arr2 = {1, 2};
        ListNode head2 = createLinkedList(arr2);
        System.out.println("\nOriginal list:");
        printLinkedList(head2);
        
        ListNode reversed2 = reverseList(head2);
        System.out.println("Reversed list:");
        printLinkedList(reversed2);
    }
}

File: cpp/ReverseLinkedList.cpp
#include <iostream>
#include <vector>
using namespace std;

/**
 * Reverse Linked List Problem Solution
 * 
 * Problem: Given the head of a singly linked list, reverse the list, 
 * and return the reversed list.
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* current = head;
        
        while (current != nullptr) {
            ListNode* nextTemp = current->next;
            current->next = prev;
            prev = current;
            current = nextTemp;
        }
        
        return prev;
    }
};

// Helper function to create linked list from array
ListNode* createLinkedList(const vector<int>& arr) {
    if (arr.empty()) return nullptr;
    
    ListNode* head = new ListNode(arr[0]);
    ListNode* current = head;
    
    for (int i = 1; i < arr.size(); i++) {
        current->next = new ListNode(arr[i]);
        current = current->next;
    }
    
    return head;
}

// Helper function to print linked list
void printLinkedList(ListNode* head) {
    ListNode* current = head;
    cout << "[";
    while (current != nullptr) {
        cout << current->val;
        if (current->next != nullptr) cout << ",";
        current = current->next;
    }
    cout << "]" << endl;
}

// Helper function to convert linked list to array
vector<int> linkedListToArray(ListNode* head) {
    vector<int> arr;
    ListNode* current = head;
    
    while (current != nullptr) {
        arr.push_back(current->val);
        current = current->next;
    }
    
    return arr;
}

// Helper function to free linked list memory
void freeLinkedList(ListNode* head) {
    ListNode* current = head;
    while (current != nullptr) {
        ListNode* temp = current;
        current = current->next;
        delete temp;
    }
}

// Main function for testing
int main() {
    Solution solution;
    
    // Test case 1
    vector<int> arr1 = {1, 2, 3, 4, 5};
    ListNode* head1 = createLinkedList(arr1);
    cout << "Original list:" << endl;
    printLinkedList(head1);
    
    ListNode* reversed1 = solution.reverseList(head1);
    cout << "Reversed list:" << endl;
    printLinkedList(reversed1);
    
    // Test case 2
    vector<int> arr2 = {1, 2};
    ListNode* head2 = createLinkedList(arr2);
    cout << "\nOriginal list:" << endl;
    printLinkedList(head2);
    
    ListNode* reversed2 = solution.reverseList(head2);
    cout << "Reversed list:" << endl;
    printLinkedList(reversed2);
    
    // Clean up memory
    freeLinkedList(reversed1);
    freeLinkedList(reversed2);
    
    return 0;
}



